examples/mimc7.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import annotations
+from chiquito.dsl import SuperCircuit, Circuit, StepType
+from chiquito.cb import eq, table
+from chiquito.util import F
+
+from mimc7_constants import ROUND_KEYS
+
+ROUNDS = 91
+
+
+# It's the best practice to wrap all values in F, even though the `assign` functions automatically wrap values in F.
+class Mimc7Constants(Circuit):
+    def setup(self):
+        self.pragma_num_steps(ROUNDS)
+        self.lookup_row = self.fixed("constant row")
+        self.lookup_c = self.fixed("constant value")
+        self.lookup_table = self.new_table(
+            table().add(self.lookup_row).add(self.lookup_c)
+        )
+
+    def fixed_gen(self):
+        for i, round_key in enumerate(ROUND_KEYS):
+            self.assign(i, self.lookup_row, F(i))
+            self.assign(i, self.lookup_c, F(round_key))
+
+
+class Mimc7Circuit(Circuit):
+    def setup(self):
+        self.x = self.forward("x")
+        self.k = self.forward("k")
+        self.c = self.forward("c")
+        self.row = self.forward("row")
+
+        self.mimc7_first_step = self.step_type(Mimc7FirstStep(self, "mimc7_first_step"))
+        self.mimc7_step = self.step_type(Mimc7Step(self, "mimc7_step"))
+        self.mimc7_last_step = self.step_type(Mimc7LastStep(self, "mimc7_last_step"))
+
+        self.pragma_first_step(self.mimc7_first_step)
+        self.pragma_last_step(self.mimc7_last_step)
+        self.pragma_num_steps(ROUNDS + 2 - 1)
+
+    def trace(self, x_in_value, k_value):
+        c_value = F(ROUND_KEYS[0])
+        x_value = F(x_in_value)
+        row_value = F(0)
+
+        self.add(self.mimc7_first_step, x_value, k_value, c_value, row_value)
+
+        for i in range(1, ROUNDS):
+            row_value += F(1)
+            x_value += F(k_value + c_value)
+            x_value = F(x_value**7)
+            c_value = F(ROUND_KEYS[i])
+
+            self.add(self.mimc7_step, x_value, k_value, c_value, row_value)
+
+        row_value += F(1)
+        x_value += F(k_value + c_value)
+        x_value = F(x_value**7)
+
+        self.add(self.mimc7_last_step, x_value, k_value, c_value, row_value)
+
+
+class Mimc7FirstStep(StepType):
```

Comment:
@qwang98 I think it is more clear to put the steps before the circuit

Reply:
Fixed.